src/snapping/index.js

```diff
@@ -133,12 +166,168 @@ class Snapping {
   enableSnapping() {
     this.snappingEnabled = true;
     this._snappableLayers().forEach((l) => this._addSnapBuffer(l));
-    this.map.on("mousemove", this._throttledMouseOverHandler);
-    this.map.on("mouseout", this._mouseoutHandler);
     this._addSnapSourceAndLayer();
+    this.map.on("mousemove", this._throttledMouseMoveHandler);
+    this.map.on("mouseout", this._mouseoutHandler);
+
+    // If the feature has been snapped to a linestring or
+    // a polygon, nearestPointOnLine may give an innacurate result (e.g., slightly off the line),
+    // especially if the line is very long. Therefore, when the vertex is "complete", we go to the
+    // database to get a point that is truly on the snapped-to feature
+    this.map.on("mousedown", async () => {
+      if (!this.snappedGeometry || !this._isLineDraw()) return;
+
+      this._handleLineStringAndPolygonSnapEnd();
+    });
+
+    this.map.on("mouseup", async () => {
+      if (!this.snappedGeometry || !this._isPointDraw()) return;
+
+      this._handlePointSnapEnd();
+    });
   }
 
-  async _mouseoverHandler(e) {
+  async _handlePointSnapEnd() {
+    if (this._isSnappedToPoint) return;
+
+    const { features } = this.store.ctx.api.getAll();
+    const feature = cloneDeep(features[0]);
+    const [lng, lat] = getCoords(feature);
+
+    const { vetro_id: vetroId } = this.snappedFeature.properties;
+
+    const closestPoint = await this.getClosestPoint(vetroId, lng, lat);
+
+    feature.geometry.coordinates = getCoord(closestPoint);
+
+    const fc = turfFeatureCollection([feature]);
+
+    this.store.ctx.api.set(fc);
+  }
+
+  async _handleLineStringAndPolygonSnapEnd() {
+    if (this._isSnappedToPoint) return;
+
+    const { features } = this.store.ctx.api.getAll();
+    const feature = cloneDeep(features[0]);
+
+    const [lng, lat] = last(getCoords(feature));
+
+    const { vetro_id: vetroId } = this.snappedFeature.properties;
+
+    const closestPoint = await this.getClosestPoint(vetroId, lng, lat);
+
+    feature.geometry.coordinates.splice(-1, 1, getCoord(closestPoint));
+
+    const fc = turfFeatureCollection([feature]);
+
+    this.store.ctx.api.set(fc);
+  }
+
+  _circleFromMousePoint(x, y) {
+    const mouseLatLng = turfPoint(this.map.unproject([x, y]).toArray());
+
+    const snapDistanceDeltaLatLng = turfPoint(
+      this.map.unproject([x + this.snapDistance, y]).toArray()
+    );
+
+    const km = turfDistance(mouseLatLng, snapDistanceDeltaLatLng);
+
+    const circle = turfCircle(this.map.unproject([x, y]).toArray(), km);
+
+    return circle;
+  }
+
+  _getClosestPoint(x, y) {
```

Comment:
**opinion**

I feel like `_getClosestPoint` is too similar to `getClosestPoint` it would be nice if the choice of function name bespoke not only their similarity, but maybe their difference as well?

Reply:
even something like `_getClosestPointSync` would just like, make it obvious which one ur looking at